libs/react/src/lib/react.tsx

```diff
@@ -175,6 +176,35 @@ export function useSubscription<T>(
   return [next, error, complete];
 }
 
+/**
+ * Given a behavior subject, and an optional project function
+ * checks if the state defined by the projection fn exists
+ * and suspends if not, by throwing a promise that resolves
+ * only once it does exist.
+ *
+ * It returns the resource.
+ *
+ * This is meant to take a stream that emits a hash table
+ * accumulating resources, eg. {1: {user: 'bob'}, 2: {user: 'sally'}}
+ * and a projection function, eg hash => hash[1]
+ *
+ * If the resource in the hash table exists, it will be returned,
+ * if not, the component will suspend.
+ *
+ * WARNING: Do not use this with mutable data, if the value at any
+ * key is replaced you will get tearing in your app when the value is
+ * sampled at different points in time by React.
+ */
+export function useResource<T, R>(
+  subject: BehaviorSubject<T>,
+  projectFn: (value: T) => R
+) {
+  if (!projectFn(subject.getValue())) {
+    throw subject.pipe(map(projectFn), take(1)).toPromise();
```

Comment:
If I understand it right, currently this code throws a promise until any **one** value is emitted (I mean to highlight `filter` vs `map` difference).

If `map(projectFn)` is nullish — it will throw a new promise again (that's interesting how suspense handles several promise throws). And the suspense might never resolve if the value is still nullish after successful resolution, e.g.: `{ 1: null, 2: 'hello world' }`.

_I might be missing something (convention for example). And I know it's a WIP._

Reply:
My bad, it's BehaviorSubject! 👀 

So that promise is going to resolve immediately and indefinitely until true-ish value is emitted? Or `React.Suspese` handles it differently? _(scratching head)_

Reply:
> If map(projectFn) is nullish — it will throw a new promise again

You're right 😨 ! Good call out! `if (!projectFn(subject.getValue())) {` returns null, undefined, false, 0, etc. will cause us to get "stuck in suspense" if the value returned by the `projectFn` never becomes truthy! I think there's a few possible solutions there

- one way is to invert control and put the impetus on the end user to do the throwing, but that would not be as ergonomic.
- another idea is to again invert control but have the user return a special `suspend` symbol in their projection function (an approach used by some "competing" React Rx libs), again not super ergonomic IMO
- another idea is to make this check more strict, and have it only throw the promise if the value is undefined (I agree this would be an improvement, with basically no down side)
- we can keep it how it is & maybe I need better docs to explain the convention?
- any other ideas I'm not considering?


I guess the main use case you bring up (always helps to frame things in terms of user stories), is that as a user I may have an API that returns `null`, and based on that information I want to unsuspend my React tree & render a "404 resource not found".

What happens if we chose `undefined` though, and the user story is that the user`s backend API also returns `undefined` which should un-suspend the tree? (I'm asking this constructively because I think we're onto something cool here, not to attack your idea, but rather to iterate on your idea)

## Some more context

The line of thinking that brought me this approach, for more context

- Looked at competing art, they subscribe in render via "magic", decided to try not to do that
- Competing art did use a custom `BehaviorObservable` which exposes a `getValue()` which causes a subscribe in render
- Rx Store is supposed to use standard RxJS concepts, so instead of following the competing art, I decided to try with `BehaviorSubject`
- I looked at the React docs and these key points stood out

> Suspense is not a data fetching library. It’s a mechanism for data fetching libraries to communicate to React that the data a component is reading is not ready yet

> It does not couple data fetching to the view layer. It helps orchestrate displaying the loading states in your UI, but it doesn’t tie your network logic to React components.

> It lets you orchestrate intentionally designed loading states. It doesn’t say how the data is fetched, but it lets you closely control the visual loading sequence of your app.

> Although it’s technically doable, Suspense is not currently intended as a way to start fetching data when a component renders. Rather, it lets components express that they’re “waiting” for data that is already being fetched

I also looked at Dan's demo app https://github.com/gaearon/suspense-experimental-github-demo as well as Relay, the implementation driving the design of React concurrent mode within React at Facebook...

[they start data fetching in the Router, not when the component mounts](https://github.com/gaearon/suspense-experimental-github-demo/blob/master/src/createRouter.js#L35 ) which runs these ["mappings" in the "data" files](https://github.com/gaearon/suspense-experimental-github-demo/blob/master/src/RepoPage.data.js) which return "resources" which [are objects that wrap a promise](https://github.com/gaearon/suspense-experimental-github-demo/blob/master/src/api.js#L33-L57). As you can see in the last link, a resource represents an immutable value that is resolved once, and throws a promise if it is not yet resolved.

This is loosely what I'm trying to mimic with RxJS, but instead of a router, I start the fetching in an Rx Store effect (outside of the React render phase). Later, when React renders, the "resource(s)" needed by the view layer are either there or not there (and it seems they can only be accessed synchronously by making them immutable by convention).

The use case I had in mind was the canonical fetching of a "resource" (JSON object) identified by some ID or key, and the way I was modeling this (by convention, for now) was a stream backed by a `BehaviorSubject` that uses `scan` operator to accumulate a hash map of immutable values that can each be represented as a promise to inform React at the point in time at which a specific "resource" (as identified by the key or projection Fn) becomes "available", with "available" being defined as anything that is "truthy", but I can see how this would cause problems for un-suspending if an API actually does return null.

Reply:
https://github.com/gaearon/suspense-experimental-github-demo/blob/master/src/api.js#L30

> // Don't copy paste this into your project!

Should we just copy paste this into our project? 🤣

Reply:
Another idea is we can have the user create the stream that does `map()` & `take(1)` (or whatever logic they want), and then wraps it in a "resource" which would basically just be copy pasting Dan's code that he said not to copy paste 🙈 . Then end users would do `resource.read()` to either get the value (which can be falsy) or throw.

I guess one tradeoff there is the "extra step" to create the resource & read it in a separate step, but its also probably more in line with the way the Relay team is doing things, so its more future proof, and would solve for the issue you bring up here.

Reply:
too many things to consider, my head is too small 🙂

**1.** Multiple promise throws:

To emphasize: I think, currently, you may get several promise throws before the value is fetched. Because `map(project).take(1).toPromise()` will resolve with the BehaviorSubject stored value in the first microtask. I suppose, this will be repeated until subject is nexted with required value. Note that with `scan` approach that might even be not the first emission — which means more time & promises will pass. Here I wanted to highlight that maybe you need to **skip** some emissions until resource is present on the subject, e.g:

```js
if (!projectFn(subject.getValue())) {
    throw subject.pipe(filter(projectFn), take(1)).toPromise();
}
```

**2.**  Falsy values:

Additional options to your suggested list:

**2.1** Key access, like "pluck"

Maybe, if you intend the resource to be `scan`-accumulated into key-value object, you could expose key-based querying, e.g:

```js
useResource(stream$, 'one');

// and under the hood something like:

useResource(subject, key) { // <- it could be (subject, ...keys) for deep object querying
  if (key in subject.getValue()) {
    throw subject.pipe(filter(v => key in v), take(1)).toPromise();
  }

  return subject.getValue()[key];
}
```

**2.2** Simplify it    
_(I lack good knowledge of rx-store atm, so it's very hypothetical. Please, don't get upset if its stupid 🙂 )_

Drop the mapping and allow only top-level store BehaviorSubjects in the useResource: 

```js
// App store creation:
root = {
  image$: new BehaviorSubject(SUSPENSE);
}

useResource<T>(subject: BehaviorSubject<T>){
  if (subject.getValue() == SUSPENSE) {
    throw ... // filter not SUSPENSE, take 1, toPromise
  }

  return subject.getValue();
}
```

And elevate the key-value resource right into the rx-store root (which is already key-based)!

**3.** Should we just copy paste this into our project? 🤣

That is the question 😅
It might complicate API and `useMemo` is not reliable (can be cleaned up) and might not be safe for concurrent mode (point made by observable-hooks creator https://twitter.com/straybugs/status/1293597197087518732 -- smart fella)

As you've already mentioned, we should discuss particular use-case scenarios and optimize on top of that 🙂

Reply:
1 - you're right, I need it to filter 👍 

2.1 - it lacks type safety, so I'd rather pass in a mapper function for that reason, but perhaps it could support both 🥇 

2.2 - this is valid feedback. I have been following this "rx store" pattern before it was "rx store" in various projects, in some apps the store value is not just string keys to subject values, but the values themselves can be arrays or objects of subjects. However I'm gravitating away from that & instead a user can always put whatever data structures they want in the subjects themselves & filter it, the advantage is it keeps it simple to iterate the available keys, the downside is maybe unnecessary filter calls *this is something I plan to revisit & doesn't have to be figured out now*. As for getting back to your feedback, you can do this now & make the mapper function `v => v` (identity function) so signify you're just waiting for any top level value. The specific use case I had in mind was caching search results by search term, or caching users by user ID, etc.

3 - I could be missing something but it seems unrelated to useMemo (whether we use Dan's `wrapPromise` utility or not?). I do agree `useMemo` is not a good place to make a subscription, which is what led me to this pattern. On that note, React has this:

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

> For example, side effects belong in useEffect, not useMemo.

What they are *NOT* very clear on is that you cannot trigger an effect based on a memoized value, which seems kind of short sighted. If the memoized value may change due to cache invalidation, that in turn will trigger any `useEffect` hooks that depend on the memoized value. *This is why we need patterns like Rx Store* - The react way has too many pitfalls, IMO. With React store, the effect(s) are started when the `<Manager />` components mount, and run until it unmounts, unless you add operators to the effect like `takeUntil`. Here's another gem for the React docs which shows we need better FRP paradigms:

https://reactjs.org/docs/concurrent-mode-suspense.html#race-conditions-with-useeffect

> if we randomize the delay time in our “fake API” implementation and press the “Next” button fast enough, we’ll see from the console logs that something is going very wrong. Requests from the previous profiles may sometimes “come back” after we’ve already switched the profile to another ID — and in that case they can overwrite the new state with a stale response for a different ID.

> This problem is possible to fix (you could use the effect cleanup function to either ignore or cancel stale requests), but it’s unintuitive and difficult to debug.

I can confirm, modeling business logic with hooks becomes very unwieldy when using hooks to model the business logic, it is why I am so in favor of something like Rx Store - where the effects are triggered only in response to `<Manager />` mounting, or specific subjects emitting events that those effects are subscribed to. There are no "gotchyas" like with React hooks. All this to say, I love React hooks! I just think that a view library can only go so far with providing good paradigms to structure business logic.

*All of this to say*, `useMemo` can be safely used to `pipe()` operators  onto streams & derive a new stream that in turn is subscribed to, but yes I think you hit the nail on the head, as soon as we memoize any source stream that produces side effects when subscribed to, we will have broken React's guarantees, so it is not only not safe to *subscribe* in useMemo, it is also not safe *to subscribe to* streams derived by a `useMemo` call (even if the subscribing happens in a `useEffect`). This is why Rx Store's effects live in the `<Manager />` component, but Rx Store probably needs better docs explaining why this tradeoff was chosen, and explaining that all side effects should be made first class Rx Store effects. I think I may even show some "invalid" usage of `useMemo` in the "nesting stores" examples, or at least I should add warnings there. So many React gotchya's 🤣  and people say RxJS is the hard one ;)

Thanks for all these ideas, keep them coming! Let me know if you're feeling ambitious enough to take on any PRs to implement any of this

Reply:
I'd love to collaborate more when I have chance to try rx-store IRL and try it's paradigms!

> Here's another gem for the React docs which shows we need better FRP paradigms

That's why I'm interested in investigating push-based framework approach — but that's a separate discussion.

---

And I'm sorry for dropping into and interrupting your WIP PR, Josh 🙂

Reply:
You're not interrupting, no worries. I appreciate any & all feedback, and just appreciate that you care enough to read the code & leave such detailed feedback! If I develop this in a vacuum, without other's feedback, it could end up just becoming a thing I understand, which is not the goal of this project which is to bring FRP to the masses of react devs :)